### PR TITLE
add SAM header fixing script to bin directory

### PR DESCRIPTION
**Pull request by rube14**
For use by P4 bam to cram new templates. This script needs a home and the bin directory in npg_seq_common  seems to be the best option.

### CODE REVIEW

**Comment by mgcam** [comment]
npg_seq_common package is for common utility modules. Historically it contains some scripts that deal with sequences. Most of this modules and scripts are being made redundant.

npg_seq_melt is a better home for this script. To be included in a suit of production scripts, this script needs further work.

File: MANIFEST [hunk 1]
```diff
@@ -21,6 +21,7 @@ bin/generate_cached_fastq
 bin/npg_fastq2fasta
 bin/npg_mail_cron_output
 bin/Ref_Maker
+bin/filter_header.pl
```
Comment (dkj) [inline MANIFEST]: Trying to encourage npg specific scripts to have a ```npg_``` prefix, so how about ```bin/npg_fix_sam_header``` ?

**Comment by dkj** [comment]
@mgcam - I'm not sure npg_seq_melt is a better place for this. Yes, we could use this functionality in merging scripts as in seq_melt, but could also use it in our standard off-instrument stuff in seq_pipeline and in custom BAM to CRAM archive conversion.

File: bin/filter_header.pl [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+use FindBin qw($Bin);
+use lib ( -d "$Bin/../lib/perl5" ? "$Bin/../lib/perl5" : "$Bin/../lib" );
+use st::api::lims;
+
+our $VERSION = '0';
+
+
+my ($rpt,$run,$pos,$tag,$af1,$af2,$lims,$sample,$library,$study,);
+
+&GetOptions("rpt:s"  => \$rpt,);
```
Comment (dkj) [inline bin/filter_header.pl]: How about a help option?

File: bin/filter_header.pl [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+# Parse file name
+if($rpt =~ /^(\d+)_(\d)(_(\w+))?(\#(\d+))?(_(\w+))?$/mxs){
```
Comment (dkj) [inline bin/filter_header.pl]: Should we use existing code to get this rpt parsing? So we only have one place to fix/update it....

File: bin/filter_header.pl [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+samtools view -H 6551_1#1.cram | filter_header.pl -rpt 6551_1#1
+
+=head1 CONFIGURATION
+
+=head1 SYNOPSIS
+
+=head1 DESCRIPTION
+
+Reads in a BAM/CRAM file header and prints out an updated version of
```
Comment (dkj) [inline bin/filter_header.pl]: So better described as a "SAM header"?
"Intended"

File: bin/filter_header.pl [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+run_position[#tag[_split]|_split].
+
+The input and output is the same: a BAM/CRAM file header in human-
+readable format (e.g. using samtools view -H).
+
+=head1 SUBROUTINES/METHODS
+
+=over
+
+=item  _get_limsm
```
Comment (dkj) [inline bin/filter_header.pl]: It's unusual to PerlDoc functions/methods marked as "private" i.e. with a ```_``` prefix - lose such PerlDoc and move the text to comments by the methods?

File: MANIFEST (hunk 1 above)
Comment (dkj) [inline MANIFEST]: MANIFEST needs fixing now script is renamed....

**Comment by dkj** [comment]
We need to use a ml_warehouse based lims driver - this is more appropriate now.

File: bin/npg_fix_sam_header [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+                  tag_index   => $tag,);
+        ($sample, $library, $study) = _get_limsi($lims);
+    }
+} elsif(defined $run && defined $pos) {
+        $lims = st::api::lims->new(
+                  driver_type => $ldt,
+                  id_run      => $run,
+                  position    => $pos,);
+        ($sample, $library, $study) = _get_limsi($lims);
+}
```
Comment (dkj) [inline bin/npg_fix_sam_header]: else croak?

File: bin/npg_fix_sam_header [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+my ($lims, $sample, $library, $study, $af);
+my $rpt = npg_tracking::glossary::rpt->inflate_rpt($rpt_key);
+my $run = $rpt->{'id_run'};
+my $pos = $rpt->{'position'};
+my $tag = $rpt->{'tag_index'};
+
+
+if (defined $run && defined $pos && defined $tag) {
+    if ($tag == 0) {
```
Comment (dkj) [inline bin/npg_fix_sam_header]: Do we need this condition? I thought st::api::lims would cope with tag_index=0

**Comment by jidur** [comment]
We would like to move this into npg_seq_melt and have the code in common between this and the reheader in iRODS script in a separate module.

**Pull request closed without merging**